table_helper.cc

```diff
@@ -129,31 +129,54 @@ future<> table_helper::setup_keyspace(cql3::query_processor& qp, service::migrat
         co_return;
     }
 
-    // FIXME: call `announce` once (collapse the calls here and in `setup_table`)
+    // FIXME: call `announce` once (`announce` keyspace and tables together)
 
     if (std::any_of(tables.begin(), tables.end(), [&] (table_helper* t) { return t->_keyspace != keyspace_name; })) {
         throw std::invalid_argument("setup_keyspace called with table_helper for different keyspace");
     }
 
     data_dictionary::database db = qp.db();
 
+    std::map<sstring, sstring> opts;
+    opts["replication_factor"] = replication_factor;
+    auto ksm = keyspace_metadata::new_keyspace(keyspace_name, "org.apache.cassandra.locator.SimpleStrategy", std::move(opts), true);
+
     if (!db.has_keyspace(keyspace_name)) {
         auto group0_guard = co_await mm.start_group0_operation();
         auto ts = group0_guard.write_timestamp();
 
         if (!db.has_keyspace(keyspace_name)) {
-            std::map<sstring, sstring> opts;
-            opts["replication_factor"] = replication_factor;
-            auto ksm = keyspace_metadata::new_keyspace(keyspace_name, "org.apache.cassandra.locator.SimpleStrategy", std::move(opts), true);
             co_await mm.announce(service::prepare_new_keyspace_announcement(db.real_database(), ksm, ts),
                     std::move(group0_guard), format("table_helper: create {} keyspace", keyspace_name));
         }
     }
 
     qs.get_client_state().set_keyspace(db.real_database(), keyspace_name);
 
-    // Create tables
-    co_await coroutine::parallel_for_each(tables, [&qp, &mm] (table_helper* t) {
-        return table_helper::setup_table(qp, mm, t->_create_cql);
```

Comment:
Is `setup_table` still used after this? If not, we could get rid of it.

Reply:
In theory, `setup_table` can be used. It is used in `table_helper::cache_table_info` to handle a case when the table doesn't exist. Additionally, it is a public function that could be used in the future if `table_helper` found a new use case.

In practice, I don't think it's used. `table_helper` is used only for `system_traces` and its tables that are created during bootstrap. So, this case in `table_helper::cache_table_info` shouldn't happen.

I'm not sure what to do with this. `table_helper` requires a big refactor (or even removal) anyway as mentioned in #15595. Maybe we can leave `setup_table` for now and think about what to do with `table_helper` later.